### PR TITLE
Extend digraph with loop detection algorithms

### PR DESCRIPTION
**Pull request by mrphrazer**
Part of this PR:
- search-first walks
- natural loop detection
- strongly connected components detection

### CODE REVIEW

File: miasm2/core/graph.py [hunk 1]
```diff
@@ -339,3 +339,167 @@ def compute_dominance_frontier(self, head):
...
+
+        # init index with 0
+        index = {v: 0 for v in self.nodes()}
+
+        for node in self.nodes():
+            # next node if node was already visited
+            if index[node]:
+                continue
+
+            todo = [(0, node)]
```
Comment (commial) [inline miasm2/core/graph.py]: `todo` elements could come from a `namedtuple`. That way, the code will be easier to understand, and less prone to error (the tuple in this list is always composed of {1,2,3}, node).

**Comment by mrphrazer** [comment]
Thanks for the feedback!

See the last commits :)

**Comment by mrphrazer** [comment]
@commial: I also don't like it that way, I misunderstood `namedtuple`.  How would you use/define the `namedtuple`, I don't see the advantage so far.

**Comment by serpilliere** [comment]
Ok. May that's commial might have said:

With `namedutple` the code transforms to:
```python
NodeState = namedtuple('NodeState', ['state', 'node'])
...
todo = [NodeState(VISIT, node)]
...
cur_node = todo.pop()
if  cur_node.state == VISIT:
    stack.append(cur_node.node)
...
todo.append(NodeState(MERGE, node))
```

**Comment by mrphrazer** [comment]
Thanks that sounds good. Fixed.

**Comment by serpilliere** [comment]
Great PR!

**Pull request merged**